cmd/zpool/zpool_iter.c

```diff
@@ -321,41 +322,259 @@ for_each_vdev(zpool_handle_t *zhp, pool_vdev_iter_f func, void *data)
+{
+	char *col = NULL;
+	char *val = line;
+	char *equals;
+	char **tmp;
+
+	if (line == NULL)
+		return (0);
+
+	equals = strchr(line, '=');
+
+	if (equals) {
+		/*
+		 * We have a 'column=value' type line.  Split it into the
+		 * column and value strings by turning the '=' into a '\0'.
+		 */
+		*equals = '\0';
+		col = line;
+		val = equals + 1;
+	} else {
+		val = line;
+	}
+
+	/* Do we already have a column by this name?  If so, skip it. */
+	if (col != NULL) {
+		for (int i = 0; i < data->cols_cnt; i++) {
+			if (strcmp(col, data->cols[i]) == 0)
+				return (1); /* Duplicate, skip */
+		}
+	}
+
+	if (val != NULL) {
+		tmp = realloc(data->lines,
+		    (data->lines_cnt + 1) * sizeof (*data->lines));
+
+		if (tmp == NULL)
+			return (0);
+
+		data->lines = tmp;
+		data->lines[data->lines_cnt] = strdup(val);
+		data->lines_cnt++;
+	}
+
+	if (col != NULL) {
+		tmp = realloc(data->cols,
+		    (data->cols_cnt + 1) * sizeof (*data->cols));
+
+		if (tmp == NULL)
+			return (0);
+
+		data->cols = tmp;
+		data->cols[data->cols_cnt] = strdup(col);
+		data->cols_cnt++;
+	}
+
+	if (val != NULL && col == NULL)
+		return (0);
+
+	return (1);
+}
+
+
+/*
+ * Run the cmd and store results in *data.
+ */
+static void
+vdev_run_cmd(vdev_cmd_data_t *data, char *cmd)
 {
-	vdev_cmd_data_t *data = cb_cmd_data;
 	char *pos = NULL;
 	FILE *fp;
 	size_t len = 0;
-	char cmd[_POSIX_ARG_MAX];
-
-	/* Set our VDEV_PATH and VDEV_UPATH env vars and run command */
-	if (snprintf(cmd, sizeof (cmd), "VDEV_PATH=%s && VDEV_UPATH=\"%s\" && "
-	    "VDEV_ENC_SYSFS_PATH=\"%s\" && %s", data->path ? data->path : "",
-	    data->upath ? data->upath : "",
-	    data->vdev_enc_sysfs_path ? data->vdev_enc_sysfs_path : "",
-	    data->cmd) >= sizeof (cmd)) {
-		/* Our string was truncated */
+	char *argv[2] = {cmd, 0};
+	char *env[5] = {"PATH=/bin:/sbin:/usr/bin:/usr/sbin", NULL, NULL, NULL,
+	    NULL};
+	char *line;
+	int link[2];
+	pid_t pid;
+	int status;
+	int rc;
+
+	if (pipe(link) == -1)
 		return;
+
+	if ((pid = fork()) == -1)
+		return; /* Couldn't fork */
+
+	if (pid == 0) {
+		/* Child process */
+		dup2(link[1], STDOUT_FILENO);
+		close(link[0]);
+		close(link[1]);
+
+		rc = asprintf(&env[1], "VDEV_PATH=%s",
+		    data->path ? data->path : "");
+		if (rc == -1)
+			goto out;
+
+		rc = asprintf(&env[2], "VDEV_UPATH=%s",
+		    data->upath ? data->upath : "");
+		if (rc == -1)
+			goto out;
+
+		rc = asprintf(&env[3], "VDEV_ENC_SYSFS_PATH=%s",
+		    data->vdev_enc_sysfs_path ?
+		    data->vdev_enc_sysfs_path : "");
+		if (rc == -1)
+			goto out;
+
+		if (env[1] != NULL && env[2] != NULL && env[3] != NULL)
+			execvpe(cmd, argv, env);
+
+	} else {
+		wait(&status);
+		close(link[1]);
+		fp = fdopen(link[0], "r");
+		if (fp == NULL)
+			return;
+
+		line = NULL;
+		/* Read back results from child process */
+		do {
+			if (getline(&line, &len, fp) != -1) {
+				if ((pos = strchr(line, '\n')) != NULL)
+					*pos = '\0';
```

Comment:
`getline()` will always return a NULL terminated line, so we're ok.  I terminated the` \n` just to keep the output formatting consistent, since scripts may or may not end the line with `\n`.